scripts/bert/pretraining_utils.py

```diff
@@ -0,0 +1,288 @@
+                                                  batch_size=batch_size * num_ctxes,
+                                                  num_buckets=num_buckets,
+                                                  ratio=0,
+                                                  shuffle=shuffle)
+            dataloader = DataLoader(dataset=dataset,
+                                    batch_sampler=sampler,
+                                    batchify_fn=batchify_fn,
+                                    num_workers=1)
+        logging.debug('Sampler created for a new dataset:\n%s', sampler.stats())
+        return dataloader
+
+    stream = stream.transform(get_dataloader)
+    return stream
+
+def get_dummy_dataloader(dataloader, target_shape):
+    """Return a dummy data loader which returns a fixed data batch of target shape"""
+    data_iter = enumerate(dataloader)
+    _, data_batch = next(data_iter)
+    logging.debug('Searching target batch shape: %s', target_shape)
+    while data_batch[0].shape != target_shape:
+        logging.debug('Skip batch with shape %s', data_batch[0].shape)
+        _, data_batch = next(data_iter)
+    logging.debug('Found target dummy batch.')
+
+    class DummyIter():
+        def __init__(self, batch):
+            self._batch = batch
+
+        def __iter__(self):
+            while True:
+                yield self._batch
+
+    return DummyIter(data_batch)
+
+def save_params(step_num, model, trainer, ckpt_dir):
+    """Save the model parameter, marked by step_num."""
+    param_path = os.path.join(ckpt_dir, '%07d.params'%step_num)
+    trainer_path = os.path.join(ckpt_dir, '%07d.states'%step_num)
+    logging.info('[step %d] Saving checkpoints to %s, %s.',
+                 step_num, param_path, trainer_path)
+    model.save_parameters(param_path)
+    trainer.save_states(trainer_path)
+
+def log(begin_time, running_num_tks, running_mlm_loss, running_nsp_loss, step_num,
+        mlm_metric, nsp_metric, trainer, log_interval):
+    """Log training progress."""
+    end_time = time.time()
+    duration = end_time - begin_time
+    throughput = running_num_tks / duration / 1000.0
+    running_mlm_loss = running_mlm_loss / log_interval
+    running_nsp_loss = running_nsp_loss / log_interval
+    lr = trainer.learning_rate if trainer else 0
+    # pylint: disable=line-too-long
+    logging.info('[step {}]\tmlm_loss={:.5f}\tmlm_acc={:.5f}\tnsp_loss={:.5f}\tnsp_acc={:.3f}\tthroughput={:.1f}K tks/s\tlr={:.7f} time={:.2f}, latency={:.1f} ms/batch'
+                 .format(step_num, running_mlm_loss.asscalar(), mlm_metric.get()[1] * 100, running_nsp_loss.asscalar(),
+                         nsp_metric.get()[1] * 100, throughput.asscalar(), lr, duration, duration*1000/log_interval))
+    # pylint: enable=line-too-long
+
+def split_and_load(arrs, ctx):
+    """split and load arrays to a list of contexts"""
+    assert isinstance(arrs, (list, tuple))
+    if len(ctx) == 1:
```

Comment:
Is it necessary to make this judgment?  This special judgment has been made in the implementation of [`mx.gluon.utils.split_and_load()`](https://mxnet.apache.org/_modules/mxnet/gluon/utils.html#split_and_load).

Reply:
Good point. removed